File: y2023/haskell/day08.hs
import Data.Map ((!))
import qualified Data.Map as Map
import qualified Text.Parsec as P

main :: IO ()
main = do
    a1
    a2

type Nodes = Map.Map String (String, String)

parser :: P.Parsec String () (String, Nodes)
parser = do
    instructions <- P.manyTill P.anyChar (P.string "\n\n")
    res <- P.many $ do
        key <- P.manyTill P.anyChar (P.string " = (")
        l   <- P.manyTill P.anyChar (P.string ", ")
        r   <- P.manyTill P.anyChar (P.string ")\n")
        return (key, (l, r))
    return (instructions, Map.fromList res)

a1 :: IO ()
a1 = do
    content <- readFile "y2023/inputs/d08"
    let Right (instr, nodes) = P.parse parser "" content
    print $ f "AAA" 0 (cycle instr) nodes
        where f "ZZZ" steps       _      _ = steps
              f  curr steps ('L':is) nodes = f (fst $ nodes ! curr) (steps+1) is nodes
              f  curr steps ('R':is) nodes = f (snd $ nodes ! curr) (steps+1) is nodes

a2 :: IO ()
a2 = do
    content <- readFile "y2023/inputs/d08"
    let Right (instr, nodes) = P.parse parser "" content
    let starts     = filter (('A'==) . last) (Map.keys nodes)
    let runF start = f start (cycle instr) nodes
    print $ foldl lcm 1 (fmap runF starts)
            where
                f curr (i:is) nodes | last curr == 'Z' = 0
                                    | i == 'L'         = 1 + f (fst $ nodes ! curr) is nodes
                                    | i == 'R'         = 1 + f (snd $ nodes ! curr) is nodes

File: y2023/kotlin/Day08.kt
package y2023

import general.Day
import general.lcm

object Day08 : Day() {
    override val name = "Haunted Wasteland"

    private fun readInput(): Pair<String, Map<String, Pair<String, String>>> {
        val lines = INPUT.readLines()
        val instructions = lines[0]
        val nodes = lines.drop(2).associate {
            val a = it.substringBefore(" ")
            val b = it.substringAfter("(").substringBefore(",")
            val c = it.substringAfter(", ").substringBefore(")")
            a to Pair(b, c)
        }
        return Pair(instructions, nodes)
    }

    override fun a1() {
        val (instructions, nodes) = readInput()

        var steps = 0
        var curr = "AAA"
        while (curr != "ZZZ") {
            val l = instructions[(steps++) % instructions.length] == 'L'
            curr = if (l) nodes[curr]!!.first else nodes[curr]!!.second
        }

        println(steps)
    }

    override fun a2() {
        val (instructions, nodes) = readInput()

        nodes.filter { it.key.endsWith("A") }
            .map {
                var steps = 0
                var curr = it.key
                while (!curr.endsWith("Z")) {
                    val l = instructions[(steps++) % instructions.length] == 'L'
                    curr = if (l) nodes[curr]!!.first else nodes[curr]!!.second
                }
                steps.toLong()
            }
            .fold(1L, ::lcm)
            .also { println(it) }
    }
}